File: baekjoon/2346/solution.py
import sys

if sys.platform != 'linux' and sys.platform != 'linux2':
  sys.stdin = open("input.txt", "r")

n = int(sys.stdin.readline().rstrip())

balloons = list(
  map(
    lambda it: int(it),
    sys.stdin.readline().rstrip().split(' ')
  )
)

result = []
length = len(balloons)
i = 0
while True:
  step = balloons[i]
  balloons[i] = None
  result.append(str(i + 1))
  if length == len(result):
    break

  while step != 0:
    i += -1 if step < 0 else 1
    if i < 0:
      i += length
    elif i >= length:
      i %= length
    if balloons[i] is not None:
      step += 1 if step < 0 else -1
    

print(' '.join(result))

File: baekjoon/2346/solution.js
function solution(n, balloons) {
  const result = [];
  const length = balloons.length;
  let i = 0;
  while (true) {
    let step = balloons[i];
    balloons[i] = null;
    result.push(i + 1);
    if (length === result.length) {
      break;
    }

    while (step !== 0) {
      i += step < 0 ? -1 : 1;
      if (i < 0) {
        i += length;
      } else if (i >= length) {
        i %= length;
      }
      if (balloons[i] !== null) {
        step += step < 0 ? 1 : -1;
      }
    }
  }

  console.log(result.join(" "));
}

//////
////
// input

const fs = require("fs");
const isLocal = process.platform !== "linux";
const filePath = isLocal ? "./input.txt" : "/dev/stdin";
const input = fs.readFileSync(filePath).toString().trim();

const mapper = (item, idx) => {
  if (idx === 0) {
    return +item;
  }

  return item
    .trim()
    .split(" ")
    .map((it) => +it);
};

if (isLocal) {
  const LINE_COUNT = 2;
  const cases = input
    .split("\n")
    .filter((item) => !!item)
    .reduce((acc, item, i) => {
      const index = Math.floor(i / LINE_COUNT);
      if (!acc[index]) {
        acc[index] = [];
      }
      acc[index].push(mapper(item, i % LINE_COUNT));

      return acc;
    }, []);

  cases.forEach((item) => {
    solution(...item);
  });
} else {
  solution(...input.split("\n").map(mapper));
}
